analyzer/process.go

```diff
@@ -1,55 +1,201 @@
 
 package analyzer
 
 import (
 	"bytes"
 	"errors"
 	"fmt"
+	"io"
 	"os"
 	"os/exec"
+	"path/filepath"
+	"strconv"
 	"strings"
 )
 
-// ErrNotFound is returned from LauncherFromPath when the zeek executable is not
-// found.
-var ErrNotFound = errors.New("executable not found")
-
-// Process is an interface for interacting running with a running process.
+// ProcessWaiter is an interface for interacting running with a process.
 type ProcessWaiter interface {
 	// Wait waits for a running process to exit, returning any errors that
 	// occur.
 	Wait() error
 }
 
-func wrapError(err error, name, stderr string) error {
+type Process struct {
+	cmd     *exec.Cmd
+	stderr  *prefixSuffixSaver
+	stdout  *prefixSuffixSaver
+	closers []io.Closer
+}
+
+func NewProcess(cmd *exec.Cmd) *Process {
+	return &Process{
+		cmd:    cmd,
+		stderr: &prefixSuffixSaver{N: 32 << 10},
+		stdout: &prefixSuffixSaver{N: 32 << 10},
+	}
+}
+
+func (p *Process) SetStdio(stderr string, stdout string) error {
+	if stderr != "" {
+		f, err := os.Create(stderr)
+		if err != nil {
+			return p.error(err)
+		}
+		p.cmd.Stderr = f
+		p.closers = append(p.closers, f)
+	}
+
+	if stdout != "" {
+		f, err := os.Create(stdout)
+		if err != nil {
+			return p.error(err)
+		}
+		p.cmd.Stdout = f
+		p.closers = append(p.closers, f)
+	}
+
+	return nil
+}
+
+func (p *Process) Start() error {
+	if p.cmd.Stderr == nil {
+		p.cmd.Stderr = p.stderr
+	} else {
+		p.cmd.Stderr = io.MultiWriter(p.stderr, p.cmd.Stderr)
+	}
+
+	if p.cmd.Stdout == nil {
+		p.cmd.Stdout = p.stdout
+	} else {
+		p.cmd.Stdout = io.MultiWriter(p.cmd.Stdout, p.stdout)
+	}
+
+	err := p.cmd.Start()
+	return p.error(err)
+}
+
+func (p *Process) Wait() error {
+	err := p.cmd.Wait()
+	for _, closer := range p.closers {
+		closer.Close()
+	}
+	return p.error(err)
+}
+
+func (p *Process) error(err error) error {
 	var exitErr *exec.ExitError
 	if errors.As(err, &exitErr) {
-		stderr = strings.TrimSpace(stderr)
-		return fmt.Errorf("%s exited with status %d: %s", name, exitErr.ExitCode(), stderr)
+		return &ProcessExitError{
+			Args:   p.cmd.Args,
+			Err:    exitErr,
+			Path:   p.cmd.Path,
+			Stderr: p.stderr.Bytes(),
+			Stdout: p.stdout.Bytes(),
+		}
 	}
-	var pathErr *os.PathError
-	if errors.As(err, &pathErr) {
-		return fmt.Errorf("error executing %s: %s: %v", name, pathErr.Path, pathErr.Err)
+	if err != nil {
+		name := filepath.Base(p.cmd.Path)
+		return fmt.Errorf("%s process error: %w", name, err)
 	}
-	return err
+
+	return nil
 }
 
-type Process struct {
-	cmd       *exec.Cmd
-	stderrBuf *bytes.Buffer
+type ProcessExitError struct {
+	Args   []string
+	Err    *exec.ExitError
+	Path   string
+	Stderr []byte
+	Stdout []byte
 }
 
-func NewProcess(cmd *exec.Cmd) *Process {
-	p := &Process{cmd: cmd, stderrBuf: bytes.NewBuffer(nil)}
-	cmd.Stderr = p.stderrBuf
-	return p
+func (p *ProcessExitError) Error() string {
+	builder := new(strings.Builder)
```

Comment:
```suggestion
	var builder strings.Builder
```

Reply:
Idk, then you have to do pointers every time you pass it into fmt.Fprint, why do you think this is better?